x/gamm/keeper/estimate_swap.go

```diff
@@ -0,0 +1,200 @@
+
+		// Execute the expected swap on the current routed pool
+		// call internal method with useCacheCtx = false to prevent double caching,
+		// we need to save pool state for the next iteration of pool route
+		tokenOutAmount, err = k.estimateSwapExactAmountIn(cacheCtx, route.PoolId, tokenIn, route.TokenOutDenom, _outMinAmount, swapFee, false)
+		if err != nil {
+			return sdk.Int{}, err
+		}
+
+		// Chain output of current pool as the input for the next routed pool
+		tokenIn = sdk.NewCoin(route.TokenOutDenom, tokenOutAmount)
+	}
+	return tokenOutAmount, nil
+}
+
+// EstimateSwapExactAmountIn estimates the amount of token out given the exact amount of token in.
+// This method does not execute the full steps of an actaul swap, nor does it change the state after estimation.
+// Returns the estimated amount of token out using cache context.
+func (k Keeper) EstimateSwapExactAmountIn(
+	ctx sdk.Context,
+	poolId uint64,
+	tokenIn sdk.Coin,
+	tokenOutDenom string,
+	tokenOutMinAmount sdk.Int,
+	swapFee sdk.Dec,
+) (sdk.Int, error) {
+	return k.estimateSwapExactAmountIn(ctx, poolId, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee, true)
+}
+
+// estimateSwapExactAmountIn returns the amount of token out given the amount of token in.
+// This method should only be called by estimate methods, as this method does not include full steps of an actual swap.
+func (k Keeper) estimateSwapExactAmountIn(
+	ctx sdk.Context,
+	poolId uint64,
+	tokenIn sdk.Coin,
+	tokenOutDenom string,
+	tokenOutMinAmount sdk.Int,
+	swapFee sdk.Dec,
+	useCacheCtx bool,
+) (sdk.Int, error) {
+	if useCacheCtx {
+		ctx, _ = ctx.CacheContext()
+	}
+	pool, err := k.getPoolForSwap(ctx, poolId)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+
+	tokenOut, err := k.swapExactAmountInNoTokenSend(ctx, pool, tokenIn, tokenOutDenom, tokenOutMinAmount, swapFee)
+	if err != nil {
+		return tokenOut.Amount, err
+	}
+
+	err = k.updatePoolForSwap(ctx, pool, tokenIn, tokenOut)
+	return tokenOut.Amount, err
+}
+
+// EstimateMultihopSwapExactAmountOut iterates `estimateSwapExactAmountOut` and returns
+// the total token in given route for multihop.
+func (k Keeper) EstimateMultihopSwapExactAmountOut(
+	ctx sdk.Context,
+	routes []types.SwapAmountOutRoute,
+	tokenInMaxAmount sdk.Int,
+	tokenOut sdk.Coin,
+) (tokenInAmount sdk.Int, err error) {
+	// use cache context so that pool state is not mutated after estimation
+	cacheCtx, _ := ctx.CacheContext()
+	swapFeeMultiplier := sdk.OneDec()
+
+	if types.SwapAmountOutRoutes(routes).IsOsmoRoutedMultihop() {
+		swapFeeMultiplier = types.MultihopSwapFeeMultiplierForOsmoPools.Clone()
+	}
+
+	// Determine what the estimated input would be for each pool along the multihop route
+	insExpected, err := k.createMultihopExpectedSwapOuts(cacheCtx, routes, tokenOut, swapFeeMultiplier)
+	if err != nil {
+		return sdk.Int{}, err
+	}
+	if len(insExpected) == 0 {
+		return sdk.Int{}, nil
+	}
+
+	insExpected[0] = tokenInMaxAmount
+
```

Comment:
Same about these needing to change with this PR https://github.com/osmosis-labs/osmosis/pull/3352

Reply:
Merged this with the most recent main and then made the according changes, can you review it and confirm?